src/main/java/frc/trigon/robot/poseestimation/relativerobotposesource/RelativeRobotPoseSource.java

```diff
@@ -0,0 +1,74 @@
+package frc.trigon.robot.poseestimation.relativerobotposesource;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform2d;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import frc.trigon.robot.RobotContainer;
+import org.littletonrobotics.junction.AutoLogOutput;
+
+/**
+ * A relative robot pose source is a robot pose source that calculates its position externally.
+ * The origin point of the relative pose source doesn't necessarily match the origin point of the robot's estimated pose, so the estimated pose needs to be transformed by that difference.
+ */
+public class RelativeRobotPoseSource {
+    private final RelativeRobotPoseSourceInputsAutoLogged inputs = new RelativeRobotPoseSourceInputsAutoLogged();
+    private final Transform2d robotCenterToCamera;
+    private final RelativeRobotPoseSourceIO relativeRobotPoseSourceIO;
+
+    private Pose2d cameraPoseAtSyncTime = new Pose2d(0, 0, new Rotation2d(0));
+    private Pose2d relativePoseSourcePoseAtSyncTime = new Pose2d(0, 0, new Rotation2d(0));
+
+    /**
+     * Constructs a new RelativeRobotPoseSource.
+     *
+     * @param robotCenterToCamera the transform of the robot's origin point to the camera
+     * @param hostname            the name of the camera in the NetworkTables
+     */
+    public RelativeRobotPoseSource(Transform2d robotCenterToCamera, String hostname) {
+        this.robotCenterToCamera = robotCenterToCamera;
+        this.relativeRobotPoseSourceIO = RelativeRobotPoseSourceIO.generateIO(hostname);
+    }
+
+    public void updatePeriodically() {
+        relativeRobotPoseSourceIO.updateInputs(inputs);
+    }
+
+    /**
+     * Resets the offset from the relative robot pose source to the robot pose if the robot isn't moving too fast.
+     * The offset is stored as two {@link Pose2d}s, one for the robot's pose and the other for the relative pose source's pose both at the same timestamp.
+     * With these two poses, we can calculate the total distance moved from the relative pose source's perspective and transform the robot's pose by the same amount.
+     *
+     * @param robotPose the current pose of the robot
+     */
+    public void resetOffset(Pose2d robotPose) {
+        if (isUnderMaximumSpeedForOffsetResetting()) {
+            this.cameraPoseAtSyncTime = robotPose.transformBy(robotCenterToCamera);
+            this.relativePoseSourcePoseAtSyncTime = inputs.pose;
+        }
+    }
+
+    @AutoLogOutput
+    public Pose2d getEstimatedRobotPose() {
+        final Transform2d movementFromSyncedPose = new Transform2d(relativePoseSourcePoseAtSyncTime, inputs.pose);
+        final Transform2d cameraToRobot = robotCenterToCamera.inverse();
+        return cameraPoseAtSyncTime.transformBy(cameraToRobot).transformBy(movementFromSyncedPose);
```

Comment:
transform the _inputs pose_ by the camToRobot (with the function mentioned earlier)